Added video processing section (Unit 7 - RNN Based Video Model) (#354)
johko/computer-vision-course: chapters/en/unit7/video-processing/rnn-based-video-models.mdx

However, they aren't designed to handle sequences where temporal relationships matter. This is where Recurrent Neural Networks (RNNs) come in. 
RNNs are specialized for processing sequential data because they have a "memory" that captures information from previous steps. This makes them well-suited for understanding how video frames relate to each other over time.

## Understanding Spatio-Temporal Modeling[[understanding-spatiotemporal-modeling]]

In video analysis, it's important to consider both spatial (space) and temporal (time) features together—this is called spatio-temporal modeling. Spatial modeling looks at what's in each frame, like objects or people, while temporal modeling looks at how these things change from frame to frame. 
By combining these two, we can understand the full context of a video. Techniques like combining CNNs and RNNs or using special types of convolutions that capture both space and time are ways researchers achieve this.

# RNN-Based Video Modeling Architectures[[rnn-based-video-modeling-architectures]]

## Long-term Recurrent Convolutional Networks(LRCN)[[lrcn]]

> <!--A image of the model architecture will be here-->

Reviewer: ```suggestion
/* A image of the model architecture will be here */
```
I would like to suggest that, according to the [mdx syntax](https://learn.microsoft.com/en-us/sql/mdx/comments-mdx-syntax?view=sql-server-ver16), comments should be handled as follows.

**Overview**
Long-term Recurrent Convolutional Networks (LRCN) are models introduced by researchers Donahue et al. in 2015. 
They combine CNNs and Long Short-Term Memory networks (LSTMs), a type of RNN, to learn from both the spatial and temporal features in videos. 
The CNN processes each frame to extract spatial features, and the LSTM takes these features in sequence to learn how they change over time.

**Key Features**
- **Combining CNN and LSTM:** Spatial features from each frame are fed into the LSTM to model the temporal relationships.
- **Versatile Applications:** LRCNs have been used successfully in tasks like action recognition (identifying actions in videos) and video captioning (generating descriptions of videos).

**Why It Matters**
LRCN was one of the first models to effectively handle both spatial and temporal aspects of video data. It paved the way for future research by showing that combining CNNs and RNNs can be powerful for video analysis.

## Convolutional LSTM Network: A Machine Learning Approach for Precipitation Nowcasting(ConvLSTM)[[convlstm]]

> <!--A image of the model architecture will be here-->

Reviewer: ```suggestion
/* A image of the model architecture will be here */
```

**Overview**

The Convolutional LSTM Network (ConvLSTM) was proposed by Shi et al. in 2015. It modifies the traditional LSTM by incorporating convolutional operations within the LSTM's structure. This means that instead of processing one-dimensional sequences, ConvLSTM can handle two-dimensional spatial data (like images) over time.

**Key Features**
- **Spatial Structure Preservation:** By using convolutions, ConvLSTM maintains the spatial layout of the data while processing temporal sequences.

> - **Effective for Spatiotemporal Prediction:** It's particularly useful for tasks that require predicting how spatial data changes over time, such as weather forecasting or video frame prediction.

Reviewer: ```suggestion
- **Effective for Spatio-Temporal Prediction:** It's particularly useful for tasks that require predicting how spatial data changes over time, such as weather forecasting or video frame prediction.
```
I would like to make a suggestion for consistency, although it is not a major change.

**Why It Matters**

> ConvLSTM introduced a new way to process spatiotemporal data by integrating convolution directly into the LSTM architecture. This has been influential in fields that need to predict future states based on spatial and temporal patterns.

Reviewer: ```suggestion
ConvLSTM introduced a new way to process spatio-temporal data by integrating convolution directly into the LSTM architecture. This has been influential in fields that need to predict future states based on spatial and temporal patterns.
```

## Unsupervised Learning of Video Representations using LSTMs[[unsupervised-lstms]]

**Overview**
In 2015, Srivastava et al. introduced a method for learning video representations without labeled data, known as unsupervised learning. This paper utilizes a multi-layer LSTM model to learn video representations. The model consists of two main components: an Encoder LSTM and a Decoder LSTM. The Encoder maps video sequences of arbitrary length (in the time dimension) to a fixed-size representation. The Decoder then uses this representation to either reconstruct the input video sequence or predict the subsequent video sequence.

**Key Features**
- **Unsupervised Learning:** The model doesn't require labeled data, making it easier to work with large amounts of video.

**Why It Matters**
This approach showed that it's possible to learn useful video representations without the need for extensive labeling, which is time-consuming and expensive. It opened up new possibilities for video analysis and generation using unsupervised methods.

## Describing Videos by Exploiting Temporal Structure[[temporal-structure]]

> <!--A image of the model architecture will be here-->

Reviewer: ```suggestion
/* A image of the model architecture will be here */
```

- **Emergence of Alternative Models**
    
    Newer models like Transformers have been developed to address some of the limitations of RNNs. Transformers use attention mechanisms to handle sequences and can process data in parallel, making them faster and more effective at capturing long-term dependencies.

# Conclusion[[conclusion]]

RNN-based models have significantly advanced the field of video analysis by providing tools to handle temporal sequences effectively. Models like LRCN, ConvLSTM, and those incorporating attention mechanisms have demonstrated the potential of combining spatial and temporal processing. However, limitations such as difficulty with long sequences, computational inefficiency, and high data requirements highlight the need for continued innovation.

Future research is likely to focus on overcoming these challenges, possibly by adopting newer architectures like Transformers, improving training efficiency, and enhancing model interpretability. These efforts aim to create models that are both powerful and practical for real-world video applications.

> ### Reference[[reference]]

Reviewer: ```suggestion
### References[[references]]
```

> # Introduction[[introduction]]

Reviewer 2: As with the other PR, I think anchors should rather be removed
```suggestion
# Introduction
```

## Videos as Sequence Data[[video-as-sequence-data]]

> Videos are made up of a series of images called frames that are played one after another to create motion. Each frame captures spatial information—the objects and scenes in the image. When these frames are shown in sequence, they also provide temporal information—how things change and move over time. 

Reviewer 2: ```suggestion
Videos are made up of a series of images called frames that are played one after another to create motion. Each frame captures spatial information — the objects and scenes in the image. When these frames are shown in sequence, they also provide temporal information — how things change and move over time. 
```

By combining these two, we can understand the full context of a video. Techniques like combining CNNs and RNNs or using special types of convolutions that capture both space and time are ways researchers achieve this.

# RNN-Based Video Modeling Architectures[[rnn-based-video-modeling-architectures]]

## Long-term Recurrent Convolutional Networks(LRCN)[[lrcn]]

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/rnn_video_models/lrcn.png" alt="LRCN model architecture">
</div>

**Overview**
Long-term Recurrent Convolutional Networks (LRCN) are models introduced by researchers Donahue et al. in 2015. 
They combine CNNs and Long Short-Term Memory networks (LSTMs), a type of RNN, to learn from both the spatial and temporal features in videos. 
The CNN processes each frame to extract spatial features, and the LSTM takes these features in sequence to learn how they change over time.

**Key Features**
- **Combining CNN and LSTM:** Spatial features from each frame are fed into the LSTM to model the temporal relationships.

> - **Versatile Applications:** LRCNs have been used successfully in tasks like action recognition (identifying actions in videos) and video captioning (generating descriptions of videos).

Reviewer 2: Do you have some examples (e.g. papers) of the applications it is used for?

Author: Sorry for the late reply. Yes, I’ve listed the links to the papers where each model was proposed in the references section at the end of the document. Do you think it would be better to place the references next to each respective model instead of at the end?

Reviewer 2: Okay, I missed the references in the end. That is actually okay for me :+1:
